### PR TITLE
Adaptive pool 2d implementation 

### PR DESCRIPTION
**Pull request by dstoiljkovicTT**
### Ticket
[27427](https://github.com/tenstorrent/tt-metal/issues/27427)

### Problem description
[Adaptive](https://docs.pytorch.org/docs/stable/generated/torch.nn.AdaptiveAvgPool2d.html) pool op is a version of pool2d operation that as an input argument gets output size. The difference between the adaptive pool and regular pool 2d is that adaptive pool can have non uniform kernel sizes and strides. So it is completely regular case that we end up with kernel widths while striding through the op such as [13,14,14,14,13]. This case as well as all the other cases that have middle kernel sizes uniform can be easily "fixed" by adding padding on one or both sides and transforming the kernel into uniform one. These cases are covered by padding the tensor and forwarding it to pool2d with changed kernel and stride values. Program factory and kernels are being reused from pool2d.

Cases that are not covered are the ones where the kernel sizes are descending or where the kernel sizes in the "middle" are not uniform.

### What's changed
Added pybindings and adequate invokes for the new operation. Added validation and padded the tensors that can be transformed as uniform. Changed the calls in models where adaptive op is needed. Added sweeps and unit tests covering the used scenarios.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17432884725)
- [x] [(Single-card) Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/runs/17432869223) one commit after the comparable main commit [ Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/runs/17400043185)
- [x] [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/17432906060)
- [x] [Nightly tt-metal L2 tests](https://github.com/tenstorrent/tt-metal/actions/runs/17432930947)

### CODE REVIEW

**Review by mbahnasTT — APPROVED** [review]

**Review by ssinghalTT — APPROVED** [review]

**Comment by stevendae** [comment]
How do you run these sweeps?

File: models/demos/yolov8s_world/tt/ttnn_yolov8s_world.py [hunk 1]
```diff
@@ -589,42 +588,20 @@ def __call__(self, x, text):
...
-                                (3, 3),
-                            )[1],
-                            tt_adaptive_to_max_pool2d(
-                                [x.shape[0], int(math.sqrt(x.shape[2])), int(math.sqrt(x.shape[2])), x.shape[3]],
-                                (3, 3),
-                            )[1],
-                        ],
-                        padding=[0, 0],
-                        dilation=[1, 1],
+                        output_size=(3, 3),
```
Comment (pavlejosipovic) [inline models/demos/yolov8s_world/tt/ttnn_yolov8s_world.py:597]: should remove tt_adaptive_to_max_pool2d

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+#include <array>
+#include <vector>
+#include <cstdint>
+
+namespace ttnn {
+namespace operations::experimental::adaptive_pool {
+
+struct AdaptiveRegion {
+    uint32_t output_h, output_w;
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp]: declara vars on a new line

File: tests/ttnn/unit_tests/operations/pool/test_adaptive_pool2d.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        (2, 512, 28, 28),
+        (1, 256, 56, 56),
+    ],
+)
+@pytest.mark.parametrize(
+    "output_size",
+    (
+        (1, 1),
+        (7, 7),
+        (14, 14),
```
Comment (pavlejosipovic) [inline tests/ttnn/unit_tests/operations/pool/test_adaptive_pool2d.py:42]: maybe add some tests that are exected to fail, since we can't map to pool2d

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pools.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+#include "adaptive_pools.hpp"
+#include "adaptive_pool_utils.hpp"
+
+#include "tt-metalium/constants.hpp"
+#include <tt-metalium/buffer_types.hpp>
+#include "ttnn/operations/conv/conv2d/conv2d_utils.hpp"
+#include "ttnn/operations/core/core.hpp"
+#include "ttnn/operations/pool/pool_utils.hpp"
+#include "ttnn/operations/pool/generic/generic_pools.hpp"
+#include "ttnn/operations/sliding_window/halo/halo.hpp"
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pools.cpp]: clean up list of includes

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pools.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+#include "ttnn/operations/sliding_window/sliding_window.hpp"
+#include "ttnn/operations/data_movement/move/move.hpp"
+#include <tt-metalium/bfloat16.hpp>
+#include <tt-metalium/math.hpp>
+namespace ttnn {
+namespace operations::experimental::adaptive_pool {
+
+using namespace ttnn::operations::pool;
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pools.cpp]: avoid using like this

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            if (!w_borders_ok) {
+                error_msg += "border elements larger than middle (padding won't help), ";
+            }
+        }
+
+        error_msg += "Correctable patterns require uniform middle kernels AND borders <= middle. ";
+        error_msg += "Input shape: (" + std::to_string(input_h) + "," + std::to_string(input_w) + "), Output shape: (" +
+                     std::to_string(output_h) + "," + std::to_string(output_w) + ").";
+
+        TT_FATAL(false, "{}", error_msg);
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.cpp]: TT_THROW in cases like this

**Review by pavlejosipovic — COMMENTED** [review]
Based on issues linked to this main issue there should be more models that should benefit from this.
Lets ask model owners if we can use this in more places?

**Review by pavlejosipovic — APPROVED** [review]

File: models/demos/yolov8s_world/tt/ttnn_yolov8s_world.py (hunk 1 above)
Comment (dstoiljkovicTT) [reply models/demos/yolov8s_world/tt/ttnn_yolov8s_world.py:597]: Removed

File: tests/ttnn/unit_tests/operations/pool/test_adaptive_pool2d.py (hunk 3 above)
Comment (dstoiljkovicTT) [reply tests/ttnn/unit_tests/operations/pool/test_adaptive_pool2d.py:42]: Added.

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.cpp (hunk 6 above)
Comment (dstoiljkovicTT) [reply ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.cpp]: Replaced

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pools.cpp (hunk 5 above)
Comment (dstoiljkovicTT) [reply ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pools.cpp]: Removed

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pools.cpp (hunk 4 above)
Comment (dstoiljkovicTT) [reply ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pools.cpp]: Done

**Comment by dstoiljkovicTT** [comment]
> How do you run these sweeps?

Since the sweeps are added to nightly folder, they are run with this pipeline [Nightly tt-metal L2 tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) for wormhole or this one [(Blackhole) Blackhole nightly tests](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-nightly-tests.yaml) for blackhole.

**Comment by tt-rkim** [comment]
If these are sweeps, perhaps migrate them to https://github.com/tenstorrent/tt-metal/actions/workflows/ttnn-run-sweeps.yaml

to take advantage of that infrastructure

**Comment by dstoiljkovicTT** [comment]
> Based on issues linked to this main issue there should be more models that should benefit from this. Lets ask model owners if we can use this in more places?

Hello @punithsekar and @keerthana-r-mcw, in this PR implementing adaptive pool op I have replaced the adaptive pool workaround call in yolov8s_world model. Could you let me know if there are other models I should replace fallback adaptive pool op call with the ttnn one? Thanks in advance

**Comment by dstoiljkovicTT** [comment]
> If these are sweeps, perhaps migrate them to https://github.com/tenstorrent/tt-metal/actions/workflows/ttnn-run-sweeps.yaml
> 
> to take advantage of that infrastructure

Okay so these tests are only tests for this op and we need them to be run daily and within nightly testing pipeline, but I see how the naming can be misleading so I changed it.

**Comment by saichandax** [comment]
> > Based on issues linked to this main issue there should be more models that should benefit from this. Lets ask model owners if we can use this in more places?
> 
> Hello @punithsekar and @keerthana-r-mcw, in this PR implementing adaptive pool op I have replaced the adaptive pool workaround call in yolov8s_world model. Could you let me know if there are other models I should replace fallback adaptive pool op call with the ttnn one? Thanks in advance

@dstoiljkovicTT ,
Workaround (ttnn.global_avg_pool2d or ttnn.avg_pool2d) is used in [Yolov9c](https://github.com/tenstorrent/tt-metal/blob/0e110b675436d358a9d428deabb4426eb3d407e3/models/demos/yolov9c/tt/ttnn_yolov9c.py#L298), [Swin_s](https://github.com/tenstorrent/tt-metal/blob/81989dcdb8f9b340c932ae7a71a346f4f08703eb/models/experimental/swin_s/tt/tt_swin_transformer.py#L115), and [EfficientNetB0](https://github.com/tenstorrent/tt-metal/blob/81989dcdb8f9b340c932ae7a71a346f4f08703eb/models/experimental/efficientnetb0/tt/efficientnetb0.py#L247).
However, replacing with AAP2d might impact performance +/-.

**Comment by dstoiljkovicTT** [comment]
> > > Based on issues linked to this main issue there should be more models that should benefit from this. Lets ask model owners if we can use this in more places?
> > 
> > 
> > Hello @punithsekar and @keerthana-r-mcw, in this PR implementing adaptive pool op I have replaced the adaptive pool workaround call in yolov8s_world model. Could you let me know if there are other models I should replace fallback adaptive pool op call with the ttnn one? Thanks in advance
> 
> @dstoiljkovicTT , Workaround (ttnn.global_avg_pool2d or ttnn.avg_pool2d) is used in [Yolov9c](https://github.com/tenstorrent/tt-metal/blob/0e110b675436d358a9d428deabb4426eb3d407e3/models/demos/yolov9c/tt/ttnn_yolov9c.py#L298), [Swin_s](https://github.com/tenstorrent/tt-metal/blob/81989dcdb8f9b340c932ae7a71a346f4f08703eb/models/experimental/swin_s/tt/tt_swin_transformer.py#L115), and [EfficientNetB0](https://github.com/tenstorrent/tt-metal/blob/81989dcdb8f9b340c932ae7a71a346f4f08703eb/models/experimental/efficientnetb0/tt/efficientnetb0.py#L247). However, replacing with AAP2d might impact performance +/-.

Yes, the global_avg_pool shouldn't be replaced since it is more performant since it goes straight to reduce. But the avg pool can be replaced since adaptive pool will not make the performance worse. So I will replace the usage of the avg_pool in Yolov9c model. Thanks

**Comment by kkeerthana0573** [comment]
@dstoiljkovicTT, 
Workaround ( ttnn.global_avg_pool2d ) is also used in Vovnet - [here](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/vovnet/tt/classifier_head.py#L24).

**Comment by dstoiljkovicTT** [comment]
> here

Thanks. I will not replace the global avg pool calls yet since those are more performant than adaptive pool.

**Comment by dstoiljkovicTT** [comment]
> @dstoiljkovicTT, Workaround ( ttnn.global_avg_pool2d ) is also used in Vovnet - [here](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/vovnet/tt/classifier_head.py#L24).

Thanks. I will not replace the global avg pool calls yet since those are more performant than adaptive pool.

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+#pragma once
+
+#include <array>
+#include <vector>
+#include <cstdint>
+
+namespace ttnn {
+namespace operations::experimental::adaptive_pool {
+
+struct AdaptiveRegion {
```
Comment (omilyutin-tt) [inline ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp]: Do we use this type?

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    std::array<uint32_t, 2> stride;
+    std::array<uint32_t, 4> padding;  // [pad_top, pad_bottom, pad_left, pad_right]
+};
+
+AdaptivePoolingParams calculate_adaptive_pool_params(
+    uint32_t input_h, uint32_t input_w, uint32_t output_h, uint32_t output_w);
+
+void validate_adaptive_pool_feasibility(uint32_t input_h, uint32_t input_w, uint32_t output_h, uint32_t output_w);
+
+// Helper functions for stride validation
```
Comment (omilyutin-tt) [inline ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp]: Hard to say what these functions do, looking only at the signature. I'd elaborate a bit more in a comment.

**Review by tt-rkim — APPROVED** [review]

**Review by stevendae — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp (hunk 7 above)
Comment (dstoiljkovicTT) [reply ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp]: No, removed, thanks.

File: ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp (hunk 8 above)
Comment (dstoiljkovicTT) [reply ttnn/cpp/ttnn/operations/experimental/adaptive_pool/adaptive_pool_utils.hpp]: Added description of each function above the definition

**Review by omilyutin-tt — APPROVED** [review]
